File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/controllers/ReportingController.java
package com.decrux.pocketr.api.controllers;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.dtos.AccountBalanceSummaryDto;
import com.decrux.pocketr.api.entities.dtos.AccountBalanceTimeseriesDto;
import com.decrux.pocketr.api.entities.dtos.MonthlyExpenseDto;
import com.decrux.pocketr.api.services.reporting.GenerateReport;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.UUID;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/v1/ledger/reports")
public class ReportingController {

    private final GenerateReport generateReport;

    public ReportingController(GenerateReport generateReport) {
        this.generateReport = generateReport;
    }

    @GetMapping("/balances")
    public List<AccountBalanceSummaryDto> getAllAccountBalances(
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate asOf,
            @AuthenticationPrincipal User user
    ) {
        return generateReport.getAllAccountBalances(user, asOf != null ? asOf : LocalDate.now());
    }

    @GetMapping("/monthly")
    public List<MonthlyExpenseDto> getMonthlyExpenses(
            @RequestParam(defaultValue = "INDIVIDUAL") String mode,
            @RequestParam @DateTimeFormat(pattern = "yyyy-MM") YearMonth period,
            @RequestParam(required = false) UUID householdId,
            @AuthenticationPrincipal User user
    ) {
        return generateReport.getMonthlyExpenses(user, period, mode, householdId);
    }

    @GetMapping("/timeseries")
    public AccountBalanceTimeseriesDto getBalanceTimeseries(
            @RequestParam UUID accountId,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate dateFrom,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate dateTo,
            @AuthenticationPrincipal User user
    ) {
        return generateReport.getBalanceTimeseries(accountId, dateFrom, dateTo, user);
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/controllers/ReportingController.kt
package com.decrux.pocketr_api.controllers

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.dtos.AccountBalanceSummaryDto
import com.decrux.pocketr_api.entities.dtos.AccountBalanceTimeseriesDto
import com.decrux.pocketr_api.entities.dtos.MonthlyExpenseDto
import com.decrux.pocketr_api.services.reporting.GenerateReport
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate
import java.time.YearMonth
import java.util.UUID

@RestController
@RequestMapping("/v1/ledger/reports")
class ReportingController(
    private val generateReport: GenerateReport,
) {
    @GetMapping("/balances")
    fun getAllAccountBalances(
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) asOf: LocalDate?,
        @AuthenticationPrincipal user: User,
    ): List<AccountBalanceSummaryDto> = generateReport.getAllAccountBalances(user, asOf ?: LocalDate.now())

    @GetMapping("/monthly")
    fun getMonthlyExpenses(
        @RequestParam(defaultValue = "INDIVIDUAL") mode: String,
        @RequestParam @DateTimeFormat(pattern = "yyyy-MM") period: YearMonth,
        @RequestParam(required = false) householdId: UUID?,
        @AuthenticationPrincipal user: User,
    ): List<MonthlyExpenseDto> = generateReport.getMonthlyExpenses(user, period, mode, householdId)

    @GetMapping("/timeseries")
    fun getBalanceTimeseries(
        @RequestParam accountId: UUID,
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) dateFrom: LocalDate,
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) dateTo: LocalDate,
        @AuthenticationPrincipal user: User,
    ): AccountBalanceTimeseriesDto = generateReport.getBalanceTimeseries(accountId, dateFrom, dateTo, user)
}
